File: 2025/day02/part02.py
def garbage(n: int, m: int):
    s = str(n)
    l = len(s)
    if l % m != 0:
        return False
    l = l // m
    base = s[0:l]
    for i in range(l, len(s), l):
        c = s[i : i + l]
        if c != base:
            return False
    return True


if __name__ == "__main__":
    with open("input.txt") as f:
        content = f.read()

    content = [elem.split("-") for elem in content.strip().split(",")]
    found = set()
    for r in content:
        a = int(r[0])
        b = int(r[1])
        for m in range(2, len(r[1]) + 1):
            for i in range(a, b + 1):
                if garbage(i, m):
                    found.add(i)

    print(sum(found))

File: 2025/day05/Part02.hs
module Part02 where

import Data.List
import Text.Read (readMaybe)

type Range = (Int, Int)

isIntersecting :: Range -> Range -> Bool
isIntersecting (al, au) (bl, bu) = let (il, iu) = (max al bl, min au bu) in il <= iu

-- mergeRange processed remaining
mergeRange :: [Range] -> [Range] -> [Range]
mergeRange a [] = a
mergeRange [] (x : xs) = mergeRange [x] xs
mergeRange processed (x : xs) =
  let (intersecting, notIntersecting) = partition (`isIntersecting` x) processed
   in let newProcessed = (minimum $ map fst $ x : intersecting, maximum $ map snd $ x : intersecting) : notIntersecting
       in mergeRange newProcessed xs

parseHeader :: String -> Maybe Range
parseHeader xs =
  case break (== '-') xs of
    (f, '-' : s) -> (,) <$> readMaybe f <*> readMaybe s
    _ -> Nothing

main :: IO ()
main = do
  content <- lines <$> readFile "input.txt"
  let (header, "" : _) = break (== "") content
  let Just header' = mapM parseHeader header
  let simplifiedRanges = mergeRange [] header'
  print $ sum $ map (\(l, u) -> u - l + 1) simplifiedRanges
  return ()
